client/allocrunner/task_hook_coordinator.go

```diff
@@ -63,15 +77,29 @@ func (c *taskHookCoordinator) setTasks(tasks []*structs.Task) {
 	if !c.hasPrestartTasks() {
 		c.mainTaskCtxCancel()
 	}
+	if !c.hasMainTasks() {
```

Comment:
Do we want to allow jobs to not have main tasks?  That seems like an invalid TaskGroup?

Also, we can close channel if there aren't any post-start tasks - this will make `taskStateUpdated` basically a no-op in the common case.